### ISSUE TITLE
Data leak in `training_GB.py`

### ISSUE THREAD

**Issue by maxall41** [AlexanderKroll/ProSmith#10]
It seems like data from the validation set is being leaked into the train set in `training_GB.py` Specifically here:
```python
dtrain_val_all_cls = xgb.DMatrix(np.concatenate([np.array(train_X_all_cls), np.array(val_X_all_cls)], axis = 0),
                                label = np.concatenate([np.array(train_labels).astype(float),np.array(val_labels).astype(float)], axis = 0))
```
Which is then used later:
```python
logging.info("Test set:")
y_test_pred_all_cls = get_predictions(param  = trials.argmin, dM_train = dtrain_val_all_cls, dM_val = dtest_all_cls)
get_performance_metrics(pred = y_test_pred_all_cls, true = test_labels)
```
Which then calls:
```python
def get_predictions(param, dM_train, dM_val):
    param, num_round, dM_train = set_param_values_V2(param = param, dtrain = dM_train)
    bst = xgb.train(param,  dM_train, num_round)
    y_val_pred = bst.predict(dM_val)
    return(y_val_pred)
```

And later `y_test_pred_all_cls` is used as part of the output of model, and the same validation set that got leaked into the training data is being used to select the optimal weighting later in the script:

```python
    best_mcc, best_mse = 0, 1000
    best_i, best_j, best_k = 0,0,0
    for i in [k/100 for k in range(0,100)]:
        for j in [k/100 for k in range(0,100)]:
            if i+j <=1:
                k = (1-i-j)
                y_val_pred = i*y_val_pred_all_cls + j*y_val_pred_all  + k*y_val_pred_cls
                if args.binary_task:
                    mcc = matthews_corrcoef(val_labels, np.round(y_val_pred))
                    if mcc > best_mcc:
                        best_mcc = mcc
                        best_i, best_j, best_k = i, j, k
                else:
                    mse = mean_squared_error(val_labels, y_val_pred)
                    if mse < best_mse:
                        best_mse = mse
                        best_i, best_j, best_k = i, j, k
y_test_pred = best_i*y_test_pred_all_cls + best_j*y_test_pred_all + best_k*y_test_pred_cls
```

Would this not result in the model performance being artificially raised?

**Comment by AlexanderKroll** [comment]
Hi Max, 

Thanks for pointing this out. There is indeed a small error. The validation set should not be included when searching for the weight parameters to combine the three models. This leads to a bit more overfitting on the validation set, but the model is ultimately evaluated on the test set, which has nothing to do with the training and validation set. So there is no data leakage from the test data and this is still a completely independent measure of model performance. It does not artificially increase model performance. 
If it does anything to the final model performance, I would expect it to be lowered because we would want to choose the weight parameters so that we get good results on independent data (which we didn't do because of the bug you pointed out). Selecting the weight parameters on the validation data without including it, and then re-training all the models using the training and validation set as the training data, might therefore further improve model performance.
